best_practices.md

```diff
@@ -0,0 +1,262 @@
+# Best Practices for C++20 Repository
+
+## 1. Pass Complex Types by Const References
+
+### Practice
+Always pass all complex types (e.g., vectors, tensors) by const references.
+
+### Explanation
+Passing complex types such as vectors and tensors by const reference avoids unnecessary copying of data. This not only enhances performance by reducing overhead but also ensures that the function does not modify the original data.
+
+### Motivation
+- **Performance**: Avoids the cost of copying large data structures.
+- **Safety**: Prevents unintended modifications to the data.
+
+### Example
+```
+void write_buffer(queue_id cq_id, Tensor& dst, std::vector<std::shared_ptr<void>> src, const std::optional<std::size_t> transfer_size = std::nullopt); // Wrong!
+void write_buffer(queue_id cq_id, Tensor& dst, const std::vector<std::shared_ptr<void>>& src, const std::optional<std::size_t> transfer_size = std::nullopt); // Right!
+```
+
+## 2. Use `std::span` for Input Parameters
+
+### Practice
+Consider using `std::span` as input instead of `std::vector`. This allows `std::array` to be used as an argument as well.
+
+### Explanation
+`std::span` is a lightweight view over a contiguous sequence of objects, such as arrays and vectors. It provides a safe and flexible way to handle array-like data structures without copying them.
+
+### Motivation
+- **Flexibility**: Enables functions to accept both `std::vector` and `std::array`.
+- **Efficiency**: Eliminates the need for copying data, enhancing performance.
+- **Safety**: Provides bounds-checked access to the underlying data.
+### Example
+```
+template <typename T>
+void print_elements(std::span<T> data) {
```

Comment:
no const &?

Reply:
Don't need it, 0 difference.

Reply:
Even worse: https://quuxplusone.github.io/blog/2021/11/09/pass-string-view-by-value/

Reply:
I was aware for string_view, but not about span. Good to know.
This slightly contradicts the "best practice" to pass by const &. Just want us to be mindful of that.